### PR TITLE
Sequential Fitting in Muon FDA interface

### LINKED ISSUE #32036: Sequential Fitting in the FDA interface
### Description
One of the WIMDA replacement requirements is to add the Sequential Fitting tab, found in Muon Analysis, to the FDA interface as well.

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR adds a Sequential Fitting tab to the FDA interface. It is one of the WIMDA replacement requirements.

**To test:**
Open FDA and load `MUSR62260`
Calculate FFT
Do some normal fitting on the Fitting tab
Then try fitting on the Sequential fitting tab

Also make sure sequential fitting still works in the Muon Analysis GUI

Fixes #32036

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by robertapplin** [comment]
I have created an issue to add an option for selecting the FFT imaginary and real parts separately #32050 . I'll do this in a separate PR once this is merged

**Review by sf1919 — CHANGES_REQUESTED** [review]
I have tested the changes and apart from a problem with FFT (which will be picked up separately in #32050 ) it functions well.

I have also reviewed the code. I have some queries about specific bits of code which would be great if you could respond to. However my biggest concern is with regards testing. You have added a huge amount of code to the basic fitting model and general fitting model but I cannot see any tests for any of the new code. Is there a reason for not testing it? 

Finally - there seems to be a bit of overlap between basic and general fitting models. Can you explain why we have these separate models? What is their role?

File: scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py [hunk 1]
```diff
@@ -787,3 +788,159 @@ def _add_workspace_to_ADS(self, workspace, name: str, directory: str) -> MuonWor
...
+            fitting_func, workspaces, parameter_values, use_initial_values)
+
+        self._update_fit_functions_after_sequential_fit(workspaces, functions)
+        self._update_fit_statuses_and_chi_squared_after_sequential_fit(workspaces, fit_statuses, chi_squared_list)
+        return functions, fit_statuses, chi_squared_list
+
+    def _do_sequential_fit(self, row_index: int, workspace_name: str, parameter_values: list, functions: list,
+                           use_initial_values: bool = False):
+        """Performs a sequential fit of the single fit data."""
+        single_function = functions[row_index - 1].clone() if not use_initial_values and row_index >= 1 else \
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:895]: When we do a sequential fit, we pass on the end parameters of the previous fit to be the initial parameters of the next fit, so I believe this is what is going on here

File: scripts/Muon/GUI/Common/fitting_widgets/general_fitting/general_fitting_model.py [hunk 2]
```diff
@@ -431,3 +424,130 @@ def _rename_workspace(self, input_name: str, workspace_name: str) -> str:
...
+
+        :param workspaces: A list of lists of workspace names e.g. [[Row 1 workspaces], [Row 2 workspaces], etc...]
+        :param parameter_values: A list of lists of parameter values e.g. [[Row 1 params], [Row 2 params], etc...]
+        :param use_initial_values: If false the parameters at the end of each fit are passed on to the next fit.
+        """
+        fitting_func = self._get_sequential_fitting_func_for_normal_fitting_mode()
+        if not self.fitting_context.simultaneous_fitting_mode:
+            workspaces = self._flatten_workspace_names(workspaces)
+        return self._perform_sequential_fit_using_func(fitting_func, workspaces, parameter_values, use_initial_values)
+
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply scripts/Muon/GUI/Common/fitting_widgets/general_fitting/general_fitting_model.py:494]: Having the same name allows the Sequential fitting tab to work with a `BasicFittingModel`, a `GeneralFittingModel` or a `TFAsymmetryFittingModel`, so this is intentional

File: scripts/Muon/GUI/Common/fitting_widgets/tf_asymmetry_fitting/tf_asymmetry_fitting_model.py [hunk 3]
```diff
@@ -722,6 +661,61 @@ def _add_fit_to_context(self, input_workspace_names: list, output_workspaces: li
...
+    def _check_tf_asymmetry_compliance(self, workspace_names: list) -> str:
+        """Checks that the workspace names provided are TF Asymmetry compliant."""
+        tf_compliant, non_compliant_names = self.check_datasets_are_tf_asymmetry_compliant(workspace_names)
+        if self.fitting_context.tf_asymmetry_mode and not tf_compliant:
+            return f"Only Groups can be fitted in TF Asymmetry mode. Please unselect the following Pairs/Diffs in " \
+                   f"the grouping tab: {non_compliant_names}"
+        else:
+            return ""
+
+    def get_all_fit_function_parameter_values_for(self, fit_function: IFunction) -> list:
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply scripts/Muon/GUI/Common/fitting_widgets/tf_asymmetry_fitting/tf_asymmetry_fitting_model.py:681]: The 'for' refers to the parameter passed into the function. So its saying get all fit function parameter values for the provided fit function. I think this is a fairly common convention used to name functions

File: scripts/test/Muon/fitting_widgets/tf_asymmetry_fitting/tf_asymmetry_fitting_model_test.py [hunk 4]
```diff
@@ -564,13 +564,15 @@ def test_that_parse_parameter_values_will_parse_the_parameters_into_normal_param
...
-        runs, groups_and_pairs = self.model.get_runs_groups_and_pairs_for_fits()
-        self.assertEqual(runs, ["62260"])
-        self.assertEqual(groups_and_pairs, ["fwd", "bwd"])
+        self.model.dataset_names = self.dataset_names
 
-        self.model._get_runs_groups_and_pairs_for_single_fit.assert_called_once_with()
+        workspace_names, runs, groups_and_pairs = self.model.get_runs_groups_and_pairs_for_fits()
+        self.assertEqual(workspace_names, self.dataset_names)
+        self.assertEqual(runs, ["20884", "20884"])
+        self.assertEqual(groups_and_pairs, ["fwd", "top"])
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply scripts/test/Muon/fitting_widgets/tf_asymmetry_fitting/tf_asymmetry_fitting_model_test.py:576]: A lot of this PR has been moving code from the TFAsymmetryModel to the General and Basic Model, so a lot of tests already exist here:
https://github.com/mantidproject/mantid/blob/9783748b9b35cb1d1f1ac24d26e6edc43d7e3fc8/scripts/test/Muon/fitting_widgets/tf_asymmetry_fitting/tf_asymmetry_fitting_model_test.py#L543

The private functions starting with `_` tend not to be tested directly, and are instead tested indirectly through public methods. https://softwareengineering.stackexchange.com/questions/100959/how-do-you-unit-test-private-methods

I've added some tests for `validate_sequential_fit` as this was not tested, thanks for spotting that

[the remark this answers is not in the document]
Comment (sf1919) [reply scripts/test/Muon/fitting_widgets/tf_asymmetry_fitting/tf_asymmetry_fitting_model_test.py:576]: If the code has moved from TFAsymmetryModel should the tests move too? Sorry - just making absolutely sure!

[the remark this answers is not in the document]
Comment (robertapplin) [reply scripts/test/Muon/fitting_widgets/tf_asymmetry_fitting/tf_asymmetry_fitting_model_test.py:576]: Yes I think some of the tests should probably move, I will try sort that out later today

**Comment by robertapplin** [comment]
@sf1919 Thanks, this should be ready for another look. I did a rebase so only the last commit is new

**Review by sf1919 — DISMISSED** [review]
Thanks for moving those tests - makes a lot more sense now.

I'm happy with the changes made and can confirm it's all still working. I have come across an issue with the Fitting tab but having checked v6.1 and a recent Nightly the erorr has crept in as part of a different PR since the last release so will raise a separate issue ticket for it.

All in all a good piece of work.

**Comment by gemmaguest** [comment]
@robertapplin There are conflicts

**Comment by robertapplin** [comment]
Thanks @gemmaguest its fixed now

**Review by sf1919 — APPROVED** [review]
Resolving merge conflict removed approval so am re-approving. Previous comments still stand and a quick test proves rebasing hasn't affected functionality.

**Pull request merged**